### PR TITLE
Add pre-commit with formatter and checkers

### PR DESCRIPTION
**Pull request by volpatto**
Hi, @allanleal and contributors!

First of all, congratulation Allan and everyone who's contributing to this project. I can see a lot of interesting applications that `autodiff` suits. The design and APIs are really nice to use!

In this PR, I'm submitting a [pre-commit](https://pre-commit.com/) inclusion for autodiff. I'm including the following hooks:

* `trailing-whitespace`: This hook trims trailing whitespace.
* `end-of-file-fixer`: Ensures that a file is either empty, or ends with one newline.
* `check-merge-conflict`: Check for files that contain merge conflict strings.
* `shellcheck`: Shell scripts conform to shellcheck.

I put the git hooks that look suitable for `autodiff` development needs. A full list of available hooks can be found [here](https://pre-commit.com/hooks.html), in case that you want to check if there is some hook that can be interesting to add.

In order to use `pre-commit`, one must install it in the environment. It's install through `conda-devev` file, so no need to worry. To enable `pre-commit`, just do in your console:

```console
$ pre-commit install
```

From now on, every commit you do will be checked and formatted. In the case that the modifications are not according to checkers and formatters, the commit will **fail**. But still no need to worry. The hooks we are using will do the modifications in-place. Just try to commit again and everything will be fine.

If you want to run the hooks even before your commit, or if you want to check the whole project, you can do the following command:

```console
$ pre-commit run --all-files
```

In order to run a specifically hook, you can do:

```console
$ pre-commit run <id> --all-files
```

where `<id>` can be, for example, `shellcheck`. In this case, we have:

```console
$ pre-commit run shellcheck --all-files
```

Therefore, only `shellcheck` will be performed.

That's all for now. If you would like to add hooks, please feel free to commit directly in this PR or tell me in a comment.

### CODE REVIEW

**Comment by volpatto** [comment]
We can add `clang-format`, but I didn't include it right now. Nonetheless, I'm including a `.clang-format` file in this PR. Should we remove it?

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+SpaceAfterTemplateKeyword: 'false'
+SpaceBeforeAssignmentOperators: 'true'
+SpaceBeforeParens: Never
+SpaceInEmptyParentheses: 'false'
+SpacesInAngles: 'false'
+SpacesInCStyleCastParentheses: 'false'
+SpacesInContainerLiterals: 'false'
+SpacesInParentheses: 'false'
+SpacesInSquareBrackets: 'false'
+Standard: Cpp11
```
Comment (supersega) [inline .clang-format]: I think we should use Cpp17 instead of Cpp11. Please correct if I am wrong.

File: environment.devenv.yml [hunk 2]
```diff
@@ -1,6 +1,7 @@
 name: autodiff
 
 dependencies:
+  - black
```
Comment (supersega) [inline environment.devenv.yml]: What does this dependency do?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -4,16 +4,13 @@ repos:
       hooks:
         - id: trailing-whitespace
         - id: end-of-file-fixer
-
-  -   repo: https://github.com/asottile/blacken-docs
-      rev: v1.3.0
-      hooks:
-        -   id: blacken-docs
+        - id: check-merge-conflict
```
Comment (supersega) [inline .pre-commit-config.yaml]: Wow, it is a very nice improvement.

File: .clang-format (hunk 1 above)
Comment (volpatto) [reply .clang-format]: Yeah, my bad! Thanks for pointing that!

File: environment.devenv.yml (hunk 2 above)
Comment (volpatto) [reply environment.devenv.yml]: This a formatter for Python codes. I thought that I removed it, this is just a leftover. Thanks!

**Comment by allanleal** [comment]
@supersega Yes, this is great addition. @volpatto is also working on GitHub Actions support to improve our CI experience. Thanks Diego!

**Comment by volpatto** [comment]
> @supersega Yes, this is great addition. @volpatto is also working on GitHub Actions support to improve our CI experience. Thanks Diego!

I'm trying the best I can! GH Actions already merged in master. Some modifications are needed, some are partially applied in #69 already. Full modifications are available in the PR #79 which I opened recently. No worries, it's really nice to contribute in this project with my 1 pence!

**Comment by volpatto** [comment]
> Thanks, It is very nice changes. I think that clang-format is a good thing to keep code clean =)
> @allanleal what you think?

@supersega just to give an update to you, I have been chatting with @allanleal and we concluded that a definition of a suitable C++ format for `autodiff` will be discussed later. Thus, in this PR, I will leave as commented code everything related to `clang-format` in `pre-commit` file. Putting it in a few words: the `.clang-format` file is still included in this PR, but it's not used at all.

So, I think that I'm done in this PR, it's ready to be merged or (more) reviewed, it's up to you all!

**Comment by allanleal** [comment]
Great - sorry Diego for the delay. This is amazing contrib. I'll merge now. What is the `shellcheck` used for?

**Comment by volpatto** [comment]
> Great - sorry Diego for the delay. This is amazing contrib. I'll merge now. What is the `shellcheck` used for?

Thanks, @allanleal! No worries. `shellcheck` will verify all `.sh` files in order to check best practices, portability, syntax and etc. There are few shell scripts in `autodiff`, but it's better have it, since some scripts are critical (for CI builds, for example).

**Comment by allanleal** [comment]
Sounds good. Thanks @volpatto .

**Pull request merged**